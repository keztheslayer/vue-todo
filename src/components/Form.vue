<template>
    <form 
        class="todo__add" 
        @submit.prevent="emitItemAdded"
    >
        <input 
            v-model="newItem"
            type="text" 
            class="todo__add-input" 
            placeholder="Type something here..." 
            required
        >
        <input 
            type="submit" 
            class="btn btn-primary btn-sm todo__add-btn" 
            value="Add" 
            :disabled="!newItem.length"
        >   
    </form>
</template>

<script>
export default {
    data() {
        return {
            newItem : '',
        };
    },
    methods : {
        emitItemAdded() {
            let item = this.newItem.toString();

            if ( item ) {
                this.$emit( 'itemAdded', item );
            }

            this.newItem = '';
        },
    },
};
</script>

<style lang="less">
.todo {
    &__add {
      width: 100%;
      display: flex;
      margin-top: 36px;

      @media (max-width: 768px){
        flex-direction: column;
      }
    }

    &__add-btn {
      border-radius: 4px;

      @media (max-width: 768px){
        margin-top: 12px;
      }
    }

    &__add-input {
      display: block;
      flex-grow: 1;
      padding: 6px 12px;
      margin-right: 6px;
      font-size: 15px;
      line-height: 1.5;
      color: #333;
      background-color: #fff;
      background-clip: padding-box;
      border: 1px solid #ced4da;
      border-radius: 4px;
      transition: border-color .15s ease-in-out,box-shadow .15s ease-in-out;

      &:focus {
        outline: none;
      }

      @media (max-width: 768px){
        margin-right: 0;
      }
    }
}
</style>
