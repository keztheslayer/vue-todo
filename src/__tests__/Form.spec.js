import Form from '../components/Form.vue';
import { mount } from '@vue/test-utils';
import Vue from 'vue';

describe( 'Form.vue', () => {
    it( 'renders correct markup', () => {
        const wrapper = mount( Form );

        expect( wrapper.element ).toMatchSnapshot();
    } );

    it( 'syncs data with input', async () => {
        const wrapper = mount( Form );
        const input = wrapper.find('.todo__add-input');

        beforeEach( () => {
            wrapper.setData(
                {
                    newTask : '',
                }
            );
        } );

        input.setValue('test');

        await Vue.nextTick();

        expect( wrapper.vm.$data.newTask ).toBe('test');

    } );

    it( 'enables add-button if newTask is set', async () => {
        const wrapper = mount( Form );
        const button = wrapper.find('.todo__add-button');

        beforeEach( () => {
            wrapper.vm.$data.newTask = '';
        } );

        wrapper.setData(
            {
                newTask : 'example',
            }
        );

        await Vue.nextTick();

        expect( button.html() ).not.toContain('disabled');

    } );

    it( 'disables add-button if newTask is empty', async () => {
        const wrapper = mount( Form );
        const button = wrapper.find('.todo__add-button');

        beforeEach( () => {
            wrapper.vm.$data.newTask = 'a random task name';
        } );

        wrapper.setData(
            {
                newTask : '',
            }
        );

        await Vue.nextTick();

        expect( button.html() ).toContain('disabled');
    } );
} );
