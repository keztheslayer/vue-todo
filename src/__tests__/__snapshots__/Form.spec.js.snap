// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form.vue renders correct markup 1`] = `
<form
  class="todo__add"
>
  <input
    class="todo__add-input"
    placeholder="Type something here..."
    required="required"
    type="text"
  />
   
  <input
    class="btn btn-primary btn-sm todo__add-button"
    disabled="disabled"
    type="submit"
    value="Add"
  />
</form>
`;
